Sources/Prometheus/Prometheus.swift

```diff
@@ -24,10 +25,11 @@ public class PrometheusClient {
     
     /// Creates prometheus formatted metrics
     ///
-    /// - Returns: Newline separated string with metrics for all Metric Trackers of this Prometheus instance
-    public func collect() -> String {
-        return self.lock.withLock {
-            return self.metrics.map { $0.collect() }.joined(separator: "\n")
+    /// - Parameters:
+    ///     - p: Promise that will succeed with a newline separated string with metrics for all Metrics this PrometheusClient handles
+    public func collect(_ p: EventLoopPromise<String>) {
+        self.lock.withLock {
+            p.succeed(result: self.metrics.map { $0.collect() }.joined(separator: "\n"))
         }
```

Comment:
I would agree with @Lukasa that "both are useful" from https://forums.swift.org/t/feedback-swift-prometheus-implementation/26928

>  Naturally having access to a String version is convenient and I don't think we should throw it away, but having both is likely helpful as well.

As I've personally "abused" the synchronous one a little bit to get "dump me all the things, I don't have an event loop and I don't care" in small toy apps. Could we keep both APIs and add a node that the blocking one should be avoided?

Maybe a note like `/// This call is BLOCKING and not intended for production use, use `collect(promise:)` instead?

WDTY @Lukasa?

Reply:
OR are we afraid that people'd abuse that call? I mean, they could always to the promise and block on it... mixed feelings if we should offer the `-> String` both answers are a bit meh :-\ Would be a no brainer if there was a "GlobalExecutionContext" but we don't have one in Swift...

Reply:
I do indeed think this will be abused. Even though it is nice for testing apps. But than still, setting up an EventLoop and creating a promise can be done in 1 line of code, so I don't think it's all that bad :)

Reply:
another option is to use a offer a callback based signature eg `collect(_ callback: (String) -> Void)`

Reply:
I think a callback based approach would be good too.

Reply:
Maybe let's do the callback instead of the promise? Then folks can wrap this to do whatever they want.

Reply:
but the callback should yield an `Result<String, Error>` or `Result<ByteBuffer, Error>` IMHO, just like the synchronous version (if we decide to have one) should be `throws -> String` / `throws -> ByteBuffer`.

Reply:
or can it truly never fail?

Reply:
Right now it can not fail. It's not throwing ATM either.

Reply:
Converting it to a callback as we speak btw :)

Reply:
where does the callback get called on actually? Should it take a `queue: DispatchQueue` or what's it executing on?

Reply:
@weissi The simplest version is the "don't ask any questions, you make sure you go where you need to". That's maximally compatible with non-dispatch-based systems, but the most annoying for Dispatch-based ones. We could pass a dispatch queue if we want.

Reply:
So much of the ecosystem is using NIO-based promises that it's nice to maintain that pattern. However, using the stdlib's `Result` and a callback would be nice.

I'm not a fan of imposing limits only through comments (regarding `/// This call is BLOCKING and not intended for production use,`) since it's almost too easy to miss that and think you have the "easy" option available.

Reply:
I'm OK with either. It's a callback right now, though using the `Result` has no real added value to me since it can never fail right now. If that ever changes it would be a SemVer Major anyways (since it'll become throwing) so we can do that later.

Reply:
That a lot of ecosystem depends on NIO Future/Promise is true, but it's not a happy place IMHO, we'll have to break all these once actual futures land or if we generalise them or whichever happens first. 

I'd be okey with NIO futures/promise since that's status quo and easy to integrate with. I do think it's a bit sad that we have to pull in NIO only for the promises but yeah, no real better option out there nowadays.

The callback is nice actually since we don't leak the NIO types then, I would like that, but I'm also okey with the `ELPromise`.

Reply:
@weissi If we ditch the promise, would that mean we can go back to using either NIO 1 or NIO 2 again? Or are there also differences in `ByteBuffer`

Reply:
`ByteBuffer`'s API differ quite a lot between 1 and 2.

Reply:
Well, there goes that hope, haha. 2 branches it is than :) Still doesn't help decide upon either Promise or normal closure though 🙁